wgsl/index.bs

```diff
@@ -4557,11 +4555,282 @@ TODO: *Stub*: Expression evaluation
 
 ## Uniformity TODO ## {#uniformity}
 
-### Uniform control flow TODO ### {#uniform-control-flow}
+### Terminology and concepts TODO ### {#uniformity-concepts}
+
+The following definitions are merely informative, trying to give an intuition for what the analysis in the next subsection is computing.
+The analysis is what actually defines these concepts, and when a program is valid or break the uniformity rules.
+
+Control-flow is said to be uniform at a point in the program if all invocations execute in lockstep at that program point.
+An expression is said to be uniform if it is in uniform control-flow, and all invocations compute the same value for it.
+A local variable is said to be uniform, if all invocations have the same value in that variable at every program point where it is live.
+
+### Uniformity analysis overview ### {#uniformity-overview}
+
+In this section we specify an analysis that verifies that functions which are only safe to call in uniform control flow (e.g. barriers and derivatives) are only called in such a context.
+
+<div class="note">Note: This analysis has the following desirable properties:
+      - Sound (meaning that it rejects every program that would break the uniformity requirements of builtins)
+      - Linear time complexity
+      - Refactoring a piece of code into a function cannot make it invalid if it was valid before
+      - If the analysis refuses a program, it provides a straightforward chain of implications that can be used to craft a good error message
+</div>
+
+The analysis analyzes each function in turn, verifying that there is a context where it is safe to call this function. It rejects the program as invalid if there is no such context.
+
+At the same time, it computies metadata about the function to help analyze its callers in turn. This means that the call graph must first be built, and functions must be analyzed from the leaves upwards, i.e. from functions that call no function outside the standard library toward the entry point. This way, whenever a function is analyzed, the metadata for all of its callees has already been computed. There is no risk of being trapped in a cycle, as recurrence is forbidden in the language.
+
+Note: another way of saying the same thing is that we do a topological sort of functions ordered by the "is a (possibly indirect) callee of" partial order, and analyze in that order.
```

Comment:
wouldn't it be easier to just say we analyze functions in the order they are specified in WGSL? Why do we need to complicate this order any more?

Reply:
We are debating allowing functions to be declared in any order. You are correct that if we do not allow calling a function before it is defined, then we can just use the program order.

Reply:
IIRC, the current consensus is to have all the top-level declarations ordered.
We can therefore describe this as just analyzing the functions in the order they are written, and add a TODO to reconsider a topological sort once the ordering is relaxed (after MVP, if ever).